sdk/eventhub/azure-eventhub/azure/eventhub/_producer.py

```diff
@@ -57,8 +65,72 @@ def _set_trace_message(event_datas, parent_span=None):
         yield ed
 
 
+class IdempotentProducerMixin(object):
+    """
+    Mixin class that provides reusable idempotent publishing related functionality shared
+    by sync and async EventHubProducer.
+
+    """
+    def _on_attach(self, source, target, properties, error):
+        # pylint: disable=protected-access, unused-argument
+        if not self._enable_idempotent_partitions or str(target) != self._target:
+            return
+        self._owner_level = properties.get(PRODUCER_EPOCH)
+        self._producer_group_id = properties.get(PRODUCER_ID)
+        self._starting_sequence_number = properties.get(PRODUCER_SEQUENCE_NUMBER)
+
+    def _set_idempotent_producer_link_configs(self):
+        self._desired_capabilities = utils.data_factory(types.AMQPArray([IDEMPOTENT_PRODUCER_SYMBOL]))
+        if self._producer_group_id is not None or\
+                self._owner_level is not None or\
+                self._starting_sequence_number is not None:
```

Comment:
If one of these is None isn't that an error?

Reply:
nope, I have tested the scenario, those three don't have to be all set at the same time.